File: examples/gen_parameters/publisher.cpp
#include <chrono>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

#include "publisher_parameters.hpp"

class Publisher : public rclcpp::Node {
 public:
  Publisher() : Node("publisher_params_node") {
    publisher_ = create_publisher<std_msgs::msg::String>("topic", 10);
    auto timer_callback = [this]() -> void {
      auto message = std_msgs::msg::String();
      // Produce reasonably large messages so we quickly reach the size limit
      // that causes the recorder to split the bag (we use this as a signal that
      // the recorder has received a sufficient number of messages).
      constexpr std::size_t kMinimumMessageSize = 8192;
      message.data = std::string(kMinimumMessageSize, 'x');
      publisher_->publish(message);
    };
    timer_ = create_wall_timer(std::chrono::milliseconds(20), timer_callback);
  }

 private:
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char* argv[]) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<Publisher>());
  rclcpp::shutdown();
  return EXIT_SUCCESS;
}

File: examples/gen_parameters/publisher.py
import rclpy
from rclpy.node import Node

from std_msgs.msg import String

from gen_parameters.publisher_parameters import publisher as publisher_parameters 


class PublisherNode(Node):
    def __init__(self):
        super().__init__("publisher_node")
        self.param_listener = publisher_parameters.ParamListener(self)
        self.params = self.param_listener.get_params()
        self.param_listener.set_user_callback(self.param_callback)
        self.print_params()

        self.name_idx = 0
        self.publisher = self.create_publisher(String, self.params.publish_topic, 10)
        publish_rate_ms = 1.0 / self.params.publish_frequency_hz
        self.timer = self.create_timer(publish_rate_ms, self.timer_callback)

    def timer_callback(self):
        name = self.params.names[self.name_idx]
        self.name_idx = (self.name_idx + 1) % len(self.params.names)
        msg = String()
        msg.data = name
        self.publisher.publish(msg)
        self.get_logger().info(f"Published: {msg.data}")

    def param_callback(self, params):
        self.params = self.param_listener.get_params()
        self.print_params()

    def print_params(self):
        self.get_logger().info(f"Current parameters:")
        self.get_logger().info(f"  names: {self.params.names}")
        self.get_logger().info(f"  publish_topic: {self.params.publish_topic}")
        self.get_logger().info(f"  publish_frequency_hz: {self.params.publish_frequency_hz}")

def main(args=None):
    rclpy.init(args=args)
    node = PublisherNode()
    rclpy.spin(node)


if __name__ == "__main__":
    main()
